application.go

```diff
@@ -261,6 +347,83 @@ EventLoop:
 	return nil
 }
 
+func (a *Application) fireMouseActions(event *tcell.EventMouse, mouseEv func(MouseAction)) {
+	atX, atY := event.Position()
+	btn := event.Buttons()
+	clickMoved := atX != int(a.mouseDownX) || atY != int(a.mouseDownY)
+	btnDiff := btn ^ a.lastMouseBtn
+
+	if atX != int(a.lastMouseX) || atY != int(a.lastMouseY) {
+		mouseEv(MouseMove)
+		a.lastMouseX = atX
+		a.lastMouseY = atY
+	}
+
+	if btnDiff&tcell.Button1 != 0 {
+		if btn&tcell.Button1 != 0 {
+			mouseEv(MouseLeftDown)
+		} else {
+			mouseEv(MouseLeftUp)
+			if !clickMoved {
+				if a.lastClickTime.Add(DoubleClickInterval).Before(time.Now()) {
+					mouseEv(MouseLeftClick)
+					a.lastClickTime = time.Now()
+				} else {
+					mouseEv(MouseLeftDoubleClick)
+					a.lastClickTime = time.Time{} // reset
+				}
+			}
+		}
+	}
+
+	if btnDiff&tcell.Button2 != 0 {
+		if btn&tcell.Button2 != 0 {
+			mouseEv(MouseMiddleDown)
+		} else {
+			mouseEv(MouseMiddleUp)
+			if !clickMoved {
+				if a.lastClickTime.Add(DoubleClickInterval).Before(time.Now()) {
+					mouseEv(MouseMiddleClick)
+					a.lastClickTime = time.Now()
+				} else {
+					mouseEv(MouseMiddleDoubleClick)
+					a.lastClickTime = time.Time{} // reset
+				}
+			}
+		}
+	}
+
+	if btnDiff&tcell.Button3 != 0 {
+		if btn&tcell.Button3 != 0 {
+			mouseEv(MouseRightDown)
+		} else {
+			mouseEv(MouseRightUp)
+			if !clickMoved {
+				if a.lastClickTime.Add(DoubleClickInterval).Before(time.Now()) {
+					mouseEv(MouseRightClick)
+					a.lastClickTime = time.Now()
+				} else {
+					mouseEv(MouseRightDoubleClick)
+					a.lastClickTime = time.Time{} // reset
+				}
+			}
+		}
+	}
+
+	if btn&tcell.WheelUp != 0 {
+		mouseEv(WheelUp)
+	}
+	if btn&tcell.WheelDown != 0 {
```

Comment:
It uses the two different WheelUp values (tcell event vs tview action), I could leave it the way it is or use something like this,

```go
	for _, x := range []struct {
		button tcell.ButtonMask
		action MouseAction
	}{
		{tcell.WheelUp, WheelUp}, {tcell.WheelDown, WheelDown},
		{tcell.WheelLeft, WheelLeft}, {tcell.WheelRight, WheelRight}} {
		if ebuttons&x.button != 0 {
			mouseEv(x.action)
		}
	}
```

Similar with the mouse down, click and double click mentioned above, however that has 4 different values rather than 2 (tcell mouse down, tview mouse down, click, double click) per button. Honestly I'd rather just have some code semi duplicated. But up to you.

Reply:
On second thought, maybe this is fine. I just pushed the changes.